tool-openssl/x509_test.cc

```diff
@@ -0,0 +1,87 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include "openssl/x509.h"
+#include <openssl/err.h>
+#include <gtest/gtest.h>
+#include <stdio.h>
+#include <string>
+#include <vector>
+#include "../tool/internal.h"
+#include "internal.h"
+
+// Test x509 -in and -out
+TEST(X509Test, X509ToolTest) {
+    std::string in_path = "test_input.der";
+    std::string out_path = "test_output.der";
+
+    X509 *x509 = X509_new();
+    ASSERT_TRUE(x509 != nullptr);
+
+    // Set validity period
+    ASSERT_TRUE(X509_gmtime_adj(X509_getm_notBefore(x509), 0));
+    ASSERT_TRUE(X509_gmtime_adj(X509_getm_notAfter(x509), 31536000L));
+
+    // Generate and set the public key
+    EVP_PKEY *pkey = EVP_PKEY_new();
+    ASSERT_TRUE(pkey != nullptr);
+    RSA *rsa = RSA_new();
+    BIGNUM *bn = BN_new();
+    ASSERT_TRUE(bn != nullptr);
+    ASSERT_TRUE(BN_set_word(bn, RSA_F4));
+    ASSERT_TRUE(RSA_generate_key_ex(rsa, 2048, bn, nullptr));
+    ASSERT_TRUE(EVP_PKEY_assign_RSA(pkey, rsa));
+    BN_free(bn);
+    ASSERT_TRUE(X509_set_pubkey(x509, pkey));
+
+    // Sign certificate
+    ASSERT_TRUE(X509_sign(x509, pkey, EVP_sha256()) > 0);
+    EVP_PKEY_free(pkey);
```

Comment:
This chunk of logic could be pulled out to a utility method for easier re-use.

Reply:
Created utility method for this logic